object.go

```diff
@@ -398,7 +398,7 @@ func (obs *obs) Put(meta *ObjectMeta, r io.Reader, opts ...ObjectOpt) (*ObjectIn
 		return nil
 	}
 
-	m, h := NewMsg(chunkSubj), sha256.New()
```

Comment:
Why is this change necessary?

Reply:
Earlier, reply field in msg was set to empty string on function exit using defer. After adding retry logic,  go test detected race between setting reply in async handler and emptying it in publish async msg. So, the defer call which emptied the reply inside msg was removed. 
The code was re-using the msg object inside the loop which called publish async msg and the test failed complaining about nats reply being not empty. Hence, the new msg object is being created in every iteration of the loop now

Reply:
Earlier, reply field in msg was set to empty string on `PublishMsgAsync` exit using defer. After adding retry logic, go test detected race between setting reply in async handler `handleAsyncReply` and emptying it in `PublishMsgAsync`. So, the defer call which emptied the reply inside PublishMsgAsync was removed. The code was re-using the msg object inside the loop which called `PublishMsgAsync` and the test failed complaining about nats reply being not empty. Hence, the new msg object is being created in every iteration of the loop now

Reply:
This is no good I think. We should not leave `Msg.Reply` set with the generated subject as it makes it impossible to re-publish by the user (as in the object store now). By the way, having `Put()` in object store recreate the message with each iteration means we have increased allocations (especially for large payloads / small chunks).

I may have an idea of how to tackle this problem - we could stop modifying the message passed by the user (which would prevent race) and utilize the low level `nc.publish()`. Here's my initial attempt from your branch: https://gist.github.com/piotrpio/796e5203e2d5be2cabda771e9ec51d54

Possibly this can be improved, but it seems to work.

Reply:
Ok but even so, with your changes the message will never be restored to its original state right? Even after the message has been processed. If we need to fix it anyway we can also improve by not modifying the message at all I think.

Reply:
I understand that a caller would assume that message would not be modified in anyway after  `PublishMsgAsync ` is called and `PubAckFuture ` is processed. This is the reason why Reply was not unset earlier. I will update the PR by applying your patch

Reply:
But it was unset before your changes: https://github.com/nats-io/nats.go/pull/1695/files#diff-176a2342c55218b70f165bed6a7731dc83d901f9513382f3c98bbd3d08ce0678L941

After `PubAckFuture` is processed, the message can be re-used by the user and thus should be in its original state:
`/ The message should not be changed *until* the PubAckFuture has been processed.`

Again, your changes are good in general and we should definitely process those but we also need to be very careful when changing any behavior in the old JetStream API.

Reply:
Yes I understand it was unset before these changes as well

Reply:
@piotrpio I have updated the PR by applying your patch with one small patch. Please review the changes